src/transformers/models/roberta/modeling_roberta.py

```diff
@@ -121,7 +122,9 @@ def forward(
         if self.position_embedding_type == "absolute":
             position_embeddings = self.position_embeddings(position_ids)
             embeddings += position_embeddings
-        embeddings = self.LayerNorm(embeddings)
+        
+        if self.normalize_embeddings:
```

Comment:
this looks correct to me

Reply:
don't we have to add the layer norm somewhere as well when `self.normalize_embeddings is False`?

Reply:
@patrickvonplaten I added a LN in the encoder part.

Here's a better overview:

Fairseq model:

```txt
RobertaHubInterface(                                                                                  
  (model): RobertaModel(                                                                                  
    (encoder): RobertaEncoder(                                                  
      (sentence_encoder): TransformerEncoder(                                                         
        (dropout_module): FairseqDropout()                                                                
        (embed_tokens): Embedding(250880, 2560, padding_idx=1)                  
        (embed_positions): LearnedPositionalEmbedding(514, 2560, padding_idx=1)                       
        (layers): ModuleList(                                                                             
          (0): TransformerEncoderLayer(                                         
            (self_attn): MultiheadAttention(                                                          
              (dropout_module): FairseqDropout()                                                          
              (k_proj): Linear(in_features=2560, out_features=2560, bias=True)  
              (v_proj): Linear(in_features=2560, out_features=2560, bias=True)                        
              (q_proj): Linear(in_features=2560, out_features=2560, bias=True)                            
              (out_proj): Linear(in_features=2560, out_features=2560, bias=True)
            )                                                                                         
            (self_attn_layer_norm): FusedLayerNorm(torch.Size([2560]), eps=1e-05, elementwise_affine=True)
            (dropout_module): FairseqDropout()                                  
            (activation_dropout_module): FairseqDropout()                                             
            (fc1): Linear(in_features=2560, out_features=10240, bias=True)                                
            (fc2): Linear(in_features=10240, out_features=2560, bias=True)      
            (final_layer_norm): FusedLayerNorm(torch.Size([2560]), eps=1e-05, elementwise_affine=True)
          )
          (35): TransformerEncoderLayer(                                                                  
            (self_attn): MultiheadAttention(                                    
              (dropout_module): FairseqDropout()                                                      
              (k_proj): Linear(in_features=2560, out_features=2560, bias=True)                            
              (v_proj): Linear(in_features=2560, out_features=2560, bias=True)  
              (q_proj): Linear(in_features=2560, out_features=2560, bias=True)                        
              (out_proj): Linear(in_features=2560, out_features=2560, bias=True)                          
            )                                                                   
            (self_attn_layer_norm): FusedLayerNorm(torch.Size([2560]), eps=1e-05, elementwise_affine=True)
            (dropout_module): FairseqDropout()                                                            
            (activation_dropout_module): FairseqDropout()                       
            (fc1): Linear(in_features=2560, out_features=10240, bias=True)                            
            (fc2): Linear(in_features=10240, out_features=2560, bias=True)                                
            (final_layer_norm): FusedLayerNorm(torch.Size([2560]), eps=1e-05, elementwise_affine=True)
          )                                                                                           
        )                                                                                                 
        (layer_norm): FusedLayerNorm(torch.Size([2560]), eps=1e-05, elementwise_affine=True)
      )                                                                                               
      (lm_head): RobertaLMHead(                                                                           
        (dense): Linear(in_features=2560, out_features=2560, bias=True)         
        (layer_norm): FusedLayerNorm(torch.Size([2560]), eps=1e-05, elementwise_affine=True)          
      )                                                                                                   
    )                                                                           
    (classification_heads): ModuleDict()                                                              
  )                                                                                                       
)
```

and the current module in Transformers:

```txt
RobertaForMaskedLM(                                                             
  (roberta): RobertaModel(                                                       
    (embeddings): RobertaEmbeddings(                                                                      
      (word_embeddings): Embedding(250880, 2560, padding_idx=1)                  
      (position_embeddings): Embedding(514, 2560, padding_idx=1)               
      (token_type_embeddings): Embedding(1, 2560)                                                         
      (LayerNorm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)      
      (dropout): Dropout(p=0.1, inplace=False)                                                        
    )                                                                          
    (encoder): RobertaEncoder(                                                   
      (layer): ModuleList(                                                                            
        (0): RobertaLayer(                                                       
          (attention): RobertaAttention(                                         
            (self): RobertaSelfAttention(                                        
              (query): Linear(in_features=2560, out_features=2560, bias=True)    
              (key): Linear(in_features=2560, out_features=2560, bias=True)      
              (value): Linear(in_features=2560, out_features=2560, bias=True)    
              (dropout): Dropout(p=0.1, inplace=False)                                                    
            )                                                                    
            (output): RobertaSelfOutput(                                         
              (dense): Linear(in_features=2560, out_features=2560, bias=True)                             
              (LayerNorm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
              (dropout): Dropout(p=0.1, inplace=False)                                                
            )                                                                  
          )                                                                      
          (intermediate): RobertaIntermediate(                                                        
            (dense): Linear(in_features=2560, out_features=10240, bias=True)     
          )                                                                    
          (output): RobertaOutput(                                               
            (dense): Linear(in_features=10240, out_features=2560, bias=True)                
            (LayerNorm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)  
            (dropout): Dropout(p=0.1, inplace=False)                           
          )                                                                      
        )
        (35): RobertaLayer(                                                     
          (attention): RobertaAttention(                                         
            (self): RobertaSelfAttention(                                                                 
              (query): Linear(in_features=2560, out_features=2560, bias=True)    
              (key): Linear(in_features=2560, out_features=2560, bias=True)    
              (value): Linear(in_features=2560, out_features=2560, bias=True)                             
              (dropout): Dropout(p=0.1, inplace=False)                         
            )                                                                                         
            (output): RobertaSelfOutput(                                       
              (dense): Linear(in_features=2560, out_features=2560, bias=True)    
              (LayerNorm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)                     
              (dropout): Dropout(p=0.1, inplace=False)                           
            )                                                                    
          )                                                                      
          (intermediate): RobertaIntermediate(                                   
            (dense): Linear(in_features=2560, out_features=10240, bias=True)     
          )                                                                      
          (output): RobertaOutput(                                                                        
            (dense): Linear(in_features=10240, out_features=2560, bias=True)     
            (LayerNorm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)  
            (dropout): Dropout(p=0.1, inplace=False)                                                      
          )                                                                      
        )                                                                                             
      )                                                                        
      (LayerNorm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)        
    )                                                                                                 
  )                                                                              
  (lm_head): RobertaLMHead(                                                    
    (dense): Linear(in_features=2560, out_features=2560, bias=True)              
    (layer_norm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)                    
    (decoder): Linear(in_features=2560, out_features=250880, bias=True)          
  )                                                                            
)   
```

Reply:
However, the model diff is `max_absolute_diff = 59.05936813354492 ` and for all tokens the output tensors are the same. Do you have any idea would could be wrong, if all tensors for different tokens/input ids yield the same values :thinking:

Reply:
I have done converting successfully in fairseq v0.10.2 , and it looks identical in transformers but different in fairseq. major change is layer norm to fused layernorm.

```
RobertaHubInterface(
  (model): RobertaModel(
    (encoder): RobertaEncoder(
      (sentence_encoder): TransformerSentenceEncoder(
        (dropout_module): FairseqDropout()
        (embed_tokens): Embedding(250880, 2560, padding_idx=1)
        (embed_positions): LearnedPositionalEmbedding(514, 2560, padding_idx=1)
        (layers): ModuleList(
          (0): TransformerSentenceEncoderLayer(
            (dropout_module): FairseqDropout()
            (activation_dropout_module): FairseqDropout()
            (self_attn): MultiheadAttention(
              (dropout_module): FairseqDropout()
              (k_proj): Linear(in_features=2560, out_features=2560, bias=True)
              (v_proj): Linear(in_features=2560, out_features=2560, bias=True)
              (q_proj): Linear(in_features=2560, out_features=2560, bias=True)
              (out_proj): Linear(in_features=2560, out_features=2560, bias=True)
            )
            (self_attn_layer_norm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
            (fc1): Linear(in_features=2560, out_features=10240, bias=True)
            (fc2): Linear(in_features=10240, out_features=2560, bias=True)
            (final_layer_norm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
          )

(35): TransformerSentenceEncoderLayer(
            (dropout_module): FairseqDropout()
            (activation_dropout_module): FairseqDropout()
            (self_attn): MultiheadAttention(
              (dropout_module): FairseqDropout()
              (k_proj): Linear(in_features=2560, out_features=2560, bias=True)
              (v_proj): Linear(in_features=2560, out_features=2560, bias=True)
              (q_proj): Linear(in_features=2560, out_features=2560, bias=True)
              (out_proj): Linear(in_features=2560, out_features=2560, bias=True)
            )
            (self_attn_layer_norm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
            (fc1): Linear(in_features=2560, out_features=10240, bias=True)
            (fc2): Linear(in_features=10240, out_features=2560, bias=True)
            (final_layer_norm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
          )
        )
        (layer_norm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
      )
      (lm_head): RobertaLMHead(
        (dense): Linear(in_features=2560, out_features=2560, bias=True)
        (layer_norm): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)
      )
    )
    (classification_heads): ModuleDict()
  )
)

Reply:
But how did you load the model with `0.10.2` :thinking: -> did you adjust the fairseq code then?

When I used `0.10.2` the following error is thrown:

```
RuntimeError: Error(s) in loading state_dict for RobertaModel:
        Missing key(s) in state_dict: "encoder.sentence_encoder.emb_layer_norm.weight", "encoder.sentence_encoder.emb_layer_norm.bias". 
        Unexpected key(s) in state_dict: "encoder.sentence_encoder.layer_norm.weight", "encoder.sentence_encoder.layer_norm.bias", "encoder.sentence_encoder.version".
```

That's the reason why I updated the conversion script to latest master (1.0.0a) version, then this importing error message disappears :D

Reply:
@stefan-it yes you're right I adjusted the fairseq code, and I remember that I encountered min absolute error of about 64 when using git cloned fairseq(1.0.0a).

Reply:
And with that code I successfully converted xlmr.xl to huggingface model. So Maybe something in fairseq 1.0.0a doesn't fit to transformers model structure.
```
torch.Size([1, 11, 250880]) torch.Size([1, 11, 250880])
max_absolute_diff = 4.482269287109375e-05
Do both models output the same tensors? 🔥
Saving model to converted_xlmr.xl2
Configuration saved in converted_xlmr.xl2/config.json
Model weights saved in converted_xlmr.xl2/pytorch_model.bin
```

Reply:
Before the layer iteration, the hidden states are idential to the fairseq model, so the error should be in the `RobertaLayer` part.

Reply:
I found out that 0.10.2 result is wrong, because both output is wrong and it makes max_absolute_diff very small.
'''
our_output
tensor([[[-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908],
         [-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908],
         [-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908],
         ...,
         [-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908],
         [-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908],
         [-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908]]],
       grad_fn=<AddBackward0>)
their_output
tensor([[[-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908],
         [-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908],
         [-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908],
         ...,
         [-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908],
         [-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908],
         [-11.1780,  -5.4424,  -7.4795,  ...,  -0.4880,   1.4000,  -4.0908]]],
       grad_fn=<AddBackward0>)
'''

Reply:
I finally managed to get it working and will push it in the next days- stay tuned:)

Reply:
@stefan-it Any update or hint for the wrong value? I'm wondering what was the cause.

Reply:
@stefan-it I managed to get the same value from farseq and pushed PR in your repo.
torch.Size([1, 10, 250880]) torch.Size([1, 10, 250880])
max_absolute_diff = 0.00022125244140614
Do both models outut the asme tensors?  🔥

Reply:
```
our_output
tensor([[[ 4.9569e+01, -1.0970e+00,  3.6279e+01,  ...,  1.3821e+00,
           1.2402e+00,  1.0905e+01],
         [ 8.5117e+00, -9.9209e-02,  3.3087e+01,  ...,  1.4223e+00,
           1.5715e+00,  1.1260e+01],
         [ 9.4228e+00,  1.8814e-01,  2.4515e+01,  ...,  2.4245e+00,
           1.0935e+00,  1.1929e+01],
         ...,
         [ 8.8886e+00, -1.7367e-02,  2.5994e+01,  ...,  1.9401e+00,
           1.8700e+00,  1.2002e+01],
         [ 9.7415e+00, -2.6768e-01,  3.2220e+01,  ...,  1.9813e+00,
           1.3128e+00,  9.6978e+00],
         [ 1.6002e+01,  1.6512e+00,  5.7907e+01,  ...,  1.9653e+00,
           1.3225e+00,  1.8848e+01]]], grad_fn=<AddBackward0>)
their_output
tensor([[[ 4.9569e+01, -1.0970e+00,  3.6280e+01,  ...,  1.3821e+00,
           1.2402e+00,  1.0905e+01],
         [ 8.5117e+00, -9.9211e-02,  3.3087e+01,  ...,  1.4223e+00,
           1.5715e+00,  1.1260e+01],
         [ 9.4228e+00,  1.8814e-01,  2.4515e+01,  ...,  2.4245e+00,
           1.0935e+00,  1.1929e+01],
         ...,
         [ 8.8886e+00, -1.7370e-02,  2.5994e+01,  ...,  1.9401e+00,
           1.8700e+00,  1.2002e+01],
         [ 9.7415e+00, -2.6768e-01,  3.2220e+01,  ...,  1.9813e+00,
           1.3128e+00,  9.6978e+00],
         [ 1.6002e+01,  1.6512e+00,  5.7907e+01,  ...,  1.9653e+00,
           1.3225e+00,  1.8848e+01]]], grad_fn=<AddBackward0>)
```